server/gc_service.go

```diff
@@ -197,3 +200,69 @@ func (s *GrpcServer) WatchGCSafePointV2(request *pdpb.WatchGCSafePointV2Request,
 		}
 	}
 }
+
+// GetAllGCSafePointV2 return all gc safe point v2.
+func (s *GrpcServer) GetAllGCSafePointV2(ctx context.Context, request *pdpb.GetAllGCSafePointV2Request) (*pdpb.GetAllGCSafePointV2Response, error) {
+	fn := func(ctx context.Context, client *grpc.ClientConn) (interface{}, error) {
+		return pdpb.NewPDClient(client).GetAllGCSafePointV2(ctx, request)
+	}
+	if rsp, err := s.unaryMiddleware(ctx, request, fn); err != nil {
+		return nil, err
+	} else if rsp != nil {
+		return rsp.(*pdpb.GetAllGCSafePointV2Response), err
+	}
+
+	startkey := endpoint.GCSafePointV2Prefix()
+	endkey := clientv3.GetPrefixRangeEnd(endpoint.GCSafePointV2Prefix())
+	_, values, revision, err := s.loadRangeFromETCD(startkey, endkey)
+
+	gcSafePoints := make([]*pdpb.GCSafePointV2, 0, len(values))
+	for _, value := range values {
+		jsonGcSafePoint := &endpoint.GCSafePointV2{}
+		if err = json.Unmarshal([]byte(value), jsonGcSafePoint); err != nil {
+			return nil, errs.ErrJSONUnmarshal.Wrap(err).GenWithStackByCause()
+		}
+		gcSafePoint := &pdpb.GCSafePointV2{
+			KeyspaceId:  jsonGcSafePoint.KeyspaceID,
+			GcSafePoint: jsonGcSafePoint.SafePoint,
+		}
+		log.Info("GetAllGCSafePointV2",
+			zap.Uint32("keyspace-id", jsonGcSafePoint.KeyspaceID),
+			zap.Uint64("gc-safe-point", jsonGcSafePoint.SafePoint))
+		gcSafePoints = append(gcSafePoints, gcSafePoint)
+	}
+
+	if err != nil {
+		return &pdpb.GetAllGCSafePointV2Response{
+			Header: s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, err.Error()),
+		}, err
+	}
+
+	return &pdpb.GetAllGCSafePointV2Response{
+		Header:       s.header(),
+		GcSafePoints: gcSafePoints,
+		Revision:     revision,
+	}, nil
+}
+
+func (s *GrpcServer) loadRangeFromETCD(startKey, endKey string) ([]string, []string, int64, error) {
```

Comment:
How about using the storage interface?

Reply:
If I move loadRangeFromETCD to storage interface, it will need to add a func to return revision,but it’s not universal for other implementations,so I implement at here.